feat(logcli): return errors from library and handle fatal in CLI

Soli0222 opened the pull request:
**What this PR does / why we need it**:
This PR refactors the logcli package to return errors instead of calling `log.Fatal`, improving the package's reusability as a library while maintaining the existing CLI behavior.

**Problems solved:**
- **Library reusability**: The logcli package can now be safely embedded in other Go applications without the risk of unexpected process termination
- **Better error handling**: Applications can implement retry logic, fallback mechanisms, and custom error handling strategies
- **Improved separation of concerns**: Business logic is decoupled from presentation/termination logic

**Key changes:**
- Modified function signatures across `pkg/logcli/` to return `error` as the last return value instead of calling `log.Fatal`
- Moved error handling to the CLI layer (`cmd/logcli/main.go`) where errors are handled with `log.Fatalf` to preserve existing behavior
- Updated 21+ functions across multiple files including:
  - `pkg/logcli/detected/fields.go`
  - `pkg/logcli/query/query.go` 
  - `pkg/logcli/print/print.go`
  - `pkg/logcli/labelquery/labels.go`
  - `pkg/logcli/seriesquery/series.go`
  - And others

**Backward compatibility:**
The CLI behavior remains completely unchanged for end users - all errors are still handled with `log.Fatalf` in the main function, preserving the existing UX while making the underlying library components reusable.

**Which issue(s) this PR fixes**:
Fixes #18754

**Special notes for your reviewer**:

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [x] Tests updated
- [x] Title matches the required conventional commits format, see [here](https://www.conventionalcommits.org/en/v1.0.0/)
  - **Note** that Promtail is considered to be feature complete, and future development for logs collection will be in [Grafana Alloy](https://github.com/grafana/alloy). As such, `feat` PRs are unlikely to be accepted unless a case can be made for the feature actually being a bug fix to existing behavior.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

Soli0222 committed:
feat(logcli): return errors from library and handle fatal in CLI

- Replace log.Fatal with error returns across pkg/logcli
- Handle errors in main.go with log.Fatalf to preserve CLI behavior
- Improve library reusability and testability without changing CLI UX

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/loki?pullRequest=18778) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/loki?pullRequest=18778) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/loki?pullRequest=18778) it.</sub>

Soli0222 commented:
I've rebased this branch on the latest `main`.

Soli0222 committed:
test(logcli): add tests for error propagation from library functions

Signed-off-by: Soli0222 <github@str08.net>

Soli0222 commented:
Hi @trevorwhitney

I've rebased this branch onto the latest `main`. The branch is conflict-free and mergeable, and all tests pass (`go test ./pkg/logcli/... ./cmd/logcli/...`). I've also added tests verifying that errors now propagate to the caller instead of terminating the process, while the CLI behavior remains unchanged.

Since this proposal was already accepted in #18754, I'd love to get it moving again — could you approve the CI workflows and take a look when you have a chance? 

Thanks!

Soli0222 committed:
fix(logcli): Return tail and part-file errors to callers

Signed-off-by: Soli0222 <github@str08.net>

Soli0222 commented on pkg/logcli/query/tail.go:
```diff
@@ -56,40 +66,50 @@ func (q *Query) TailQuery(delayFor time.Duration, c client.Client, out output.Lo
+						return nil
+					}
+					log.Println("error recreating tailing connection:", err)
+					return nil
```
Fixed in 41f0a6e. Both the reconnection-exhausted path and the unrecoverable read-error path now return the wrapped error so the caller exits non-zero. The `ctx.Err() != nil` guards are kept so a user-initiated shutdown still returns nil.

Soli0222 commented on pkg/logcli/query/query.go:
```diff
@@ -220,27 +230,27 @@ func (q *Query) outputFilename() string {
 		exists, err := partFile.Exists()
 		if err != nil {
-			log.Fatalf("Query failed: %s\n", err)
+			return nil, false, fmt.Errorf("query failed: %w", err)
```
Fixed in 41f0a6e. Changed to `"checking part file existence: %w"` and `"creating part file: %w"` since the caller already wraps with "Query failed:".

Soli0222 commented on pkg/logcli/query/query.go:
```diff
@@ -175,11 +184,11 @@ func (q *Query) DoQuery(c client.Client, out output.LogOutput, statistics bool)
 					"the same timestamp, so when a batch ends in this scenario the next query will include "+
 					"all the overlapping entries again).  Please increase your batch size to at least %v to account "+
-					"for overlapping entryes\n", q.BatchSize, len(lastEntry), len(lastEntry)+1)
+					"for overlapping entryes", q.BatchSize, len(lastEntry), len(lastEntry)+1)
```
Fixed in 41f0a6e.

Soli0222 commented on pkg/logcli/query/query_test.go:
```diff
@@ -409,6 +409,17 @@ func Test_batch(t *testing.T) {
 type testQueryClient struct {
```
Fixed in 41f0a6e with `require.NoError(t, q.DoQuery(...))`.

Soli0222 commented on pkg/logcli/query/tail.go:
```diff
@@ -115,13 +135,15 @@ func (q *Query) TailQuery(delayFor time.Duration, c client.Client, out output.Lo
 
 			for _, entry := range stream.Entries {
-				out.FormatAndPrintln(entry.Timestamp, labels, 0, entry.Line)
+				if err := out.FormatAndPrintln(entry.Timestamp, labels, 0, entry.Line); err != nil {
```
This is a real bug (`q.ShowLabelsKey` should be `q.IgnoreLabelsKey`), but it predates this PR and is unrelated to error propagation, so leaving it for a separate fix.

Soli0222 commented on pkg/logcli/output/output.go:
```diff
@@ -28,7 +28,7 @@ var colorList = []*color.Color{
 // LogOutput is the interface any output mode must implement
 type LogOutput interface {
-	FormatAndPrintln(ts time.Time, lbls loghttp.LabelSet, maxLabelsLen int, line string)
+	FormatAndPrintln(ts time.Time, lbls loghttp.LabelSet, maxLabelsLen int, line string) error
```
Discarding the return value is valid Go and compiles fine, but updated the tests to use `require.NoError` in 41f0a6e anyway.

Soli0222 commented:
@matyax @trevorwhitney Thanks for the review. I've addressed the Copilot feedback in 41f0a6ea79ab286e85e94ff83e6e08b32cb250c7:

- `TailQuery` now returns the wrapped error from the reconnection-exhausted and unrecoverable read-error paths, so the CLI exits non-zero instead of silently returning. The `ctx.Err()` guards are kept so a user-initiated shutdown (Ctrl-C) still exits cleanly.
- Removed the duplicated `"query failed"` wrapping in `createPartFile`; the errors now describe the actual operation (`checking part file existence` / `creating part file`).
- Test call sites now assert the returned errors (`require.NoError` for `DoQuery` and `FormatAndPrintln`).
- Fixed the `entryes` typo.

One comment I left unresolved: the `IgnoreLabelsKey` branch in `TailQuery` passing `q.ShowLabelsKey` is indeed a real bug, but it predates this PR and is unrelated to error propagation, so I'd prefer to address it in a separate fix. Let me know if you'd rather have it fixed here.

Soli0222 committed:
fix(logcli): Handle tail cancellation and parallel errors safely

Signed-off-by: Soli0222 <github@str08.net>

Soli0222 committed:
fix(logcli): Clean up tail connections safely

Signed-off-by: Soli0222 <github@str08.net>

Soli0222 committed:
fix(logcli): Make tail reconnection cancelable

Signed-off-by: Soli0222 <github@str08.net>

Soli0222 commented:
@matyax @trevorwhitney Gentle follow-up on this PR.

I’ve now addressed all in-scope feedback from both Copilot and Cursor Bugbot. The latest fixes cover tail cancellation, WebSocket cleanup, error propagation, and cancellation during reconnects.

The relevant client and query package tests pass locally, including the race-detector tests, and Cursor Bugbot is passing. The PR currently has no merge conflicts.

However, 10 GitHub Actions workflows are still awaiting maintainer approval, so the full CI suite has not run yet.

The remaining Copilot thread concerns a pre-existing IgnoreLabelsKey bug that is unrelated to this error-handling change, so I’ve intentionally left it out of this PR.

Could someone from the team please approve the workflows and take a final human review, or assign the appropriate reviewer? This PR has been open since August 2025, and I’m happy to address any remaining
maintainer feedback.

Thanks!

chaudum commented on cmd/logcli/main.go:
```diff
@@ -479,17 +485,25 @@ func main() {
 	case labelsCmd.FullCommand():
-		labelsQuery.DoLabels(queryClient)
+		if err := labelsQuery.DoLabels(queryClient); err != nil {
+			log.Fatalf("labels failed: %s", err)
```
We should be consistent with uppercase/lowercase error messages:

```suggestion
			log.Fatalf("Labels query failed: %s", err)
```

chaudum commented on cmd/logcli/main.go:
```diff
@@ -479,17 +485,25 @@ func main() {
 	case statsCmd.FullCommand():
-		statsQuery.DoStats(queryClient)
+		if err := statsQuery.DoStats(queryClient); err != nil {
+			log.Fatalf("stats failed: %s", err)
```
```suggestion
			log.Fatalf("Stats query failed: %s", err)
```

chaudum commented on cmd/logcli/main.go:
```diff
@@ -479,17 +485,25 @@ func main() {
 	case seriesCmd.FullCommand():
-		seriesQuery.DoSeries(queryClient)
+		if err := seriesQuery.DoSeries(queryClient); err != nil {
+			log.Fatalf("series failed: %s", err)
```
```suggestion
			log.Fatalf("Series query failed: %s", err)
```

chaudum commented on cmd/logcli/main.go:
```diff
@@ -479,17 +485,25 @@ func main() {
 
-		instantQuery.DoQuery(queryClient, out, *statistics)
+		if err := instantQuery.DoQuery(queryClient, out, *statistics); err != nil {
+			log.Fatalf("Query failed: %s", err)
```
```suggestion
			log.Fatalf("Instant query failed: %s", err)
```

chaudum commented on cmd/logcli/main.go:
```diff
@@ -454,13 +454,19 @@ func main() {
 			rangeQuery.Limit = 0
-			rangeQuery.DoQueryParallel(queryClient, out, *statistics)
+			if err := rangeQuery.DoQueryParallel(queryClient, out, *statistics); err != nil {
+				log.Fatalf("Query failed: %s", err)
```
```suggestion
				log.Fatalf("Range query failed: %s", err)
```

chaudum commented on cmd/logcli/main.go:
```diff
@@ -454,13 +454,19 @@ func main() {
 		} else if rangeQuery.ParallelMaxWorkers == 1 {
-			rangeQuery.DoQuery(queryClient, out, *statistics)
+			if err := rangeQuery.DoQuery(queryClient, out, *statistics); err != nil {
+				log.Fatalf("Query failed: %s", err)
```
```suggestion
				log.Fatalf("Range query failed: %s", err)
```

chaudum commented on cmd/logcli/main.go:
```diff
@@ -508,12 +522,18 @@ func main() {
 		if cmd == volumeRangeCmd.FullCommand() {
-			index.GetVolumeRange(volumeRangeQuery, queryClient, out, *statistics)
+			if err := index.GetVolumeRange(volumeRangeQuery, queryClient, out, *statistics); err != nil {
+				log.Fatalf("volume_range failed: %s", err)
```
```suggestion
				log.Fatalf("Volume range query failed: %s", err)
```

chaudum commented on cmd/logcli/main.go:
```diff
@@ -508,12 +522,18 @@ func main() {
 		} else {
-			index.GetVolume(volumeQuery, queryClient, out, *statistics)
+			if err := index.GetVolume(volumeQuery, queryClient, out, *statistics); err != nil {
+				log.Fatalf("volume failed: %s", err)
```
```suggestion
				log.Fatalf("Volume query failed: %s", err)
```

chaudum commented on cmd/logcli/main.go:
```diff
@@ -508,12 +522,18 @@ func main() {
 	case detectedFieldsCmd.FullCommand():
-		detectedFieldsQuery.Do(queryClient, *outputMode)
+		if err := detectedFieldsQuery.Do(queryClient, *outputMode); err != nil {
+			log.Fatalf("detected-fields failed: %s", err)
```
```suggestion
			log.Fatalf("Detected-fields query failed: %s", err)
```

bboreham commented:
This PR seems to do way more than is claimed in the title and description. "Refactor" means to restructure without changing behaviour.
For instance 120 lines of code are added to `pkg/logcli/query/tail.go`; I can see no mention of "tail" in your PR description.

It is essential that you demonstrate understanding of the changes you are making.  Please ensure the description is complete, and split the PR down into separate concerns if appropriate.

chaudum reviewed (commented):
We should be consistent with the uppercase/lowercase messages returned on error.

Otherwise the changes look reasonable.

I would have liked to see the internal changes in the tail query code, which are unrelated to returning an error as it seems, in a separate PR. That would be great. However, I also reviewed that part and it also looks fine to me.

Soli0222 commented:
Thanks for the reviews, @bboreham and @chaudum — both of you were right that this had grown beyond what the title and description claimed. I've reworked the PR:

**Removed the out-of-scope tail changes.** The tail connection lifecycle work (cancellation plumbing, WebSocket cleanup, context-aware reconnection, and the `LiveTailQueryConnContext` client method) is gone from this PR. I'll open it as a follow-up once this lands. `pkg/logcli/query/tail.go` is down from 174 changed lines to 43, and the PR as a whole from 26 files to 23.

**Applied the capitalization suggestions** — thanks @chaudum. While doing that I noticed I had also silently lowercased nine log messages in `tail.go`, which was both unrelated to error propagation and inconsistent with the direction of your suggestions, so I reverted those too. There are now no `log.Print*` message changes left in the diff.

**Rewrote the description** to list every changed signature, and to call out the three intentional behavior changes explicitly rather than hiding them behind the word "refactor":

1. Tail read failures and exhausted reconnects now return an error, so `logcli tail` exits 1 instead of 0.
2. `TailQuery` no longer calls `os.Exit(0)` from its signal handler — it cancels an internal context so the loop returns normally. This one is the core of the issue: `os.Exit` in library code kills the host process.
3. `output.LogOutput.FormatAndPrintln` now returns `error`, which is breaking for out-of-tree implementations.

The branch is rebased onto the latest `main` and `go build` / `go test ./pkg/logcli/... ./cmd/logcli/...` pass locally. The GitHub Actions workflows are still awaiting maintainer approval, so the full CI suite hasn't run yet — could one of you approve them?
